### PR TITLE
add the matrix exponential, `expm`

### PR DESCRIPTION
**Pull request by ev-br**
This is split off gh-7660, per https://github.com/cupy/cupy/pull/7660#discussion_r1270266633.

For the moment, this only implements the Higham'2005 algorithm [1], in a simplified form:
- [13/13] Pade approximant only (no selection of orders 3-5-7-9)
- no matrix balancing (this requires LAPACK's *gebal which is not available on CUDA): balancing could maybe improve numerical stability.
- the implementation is pure python + matmuls.

This implementation has lots of possible improvements:
- move matmuls to cuBLAS calls, cf https://github.com/cupy/cupy/pull/7660#discussion_r1270262702 
- algorithmic improvements: cf https://github.com/cupy/cupy/pull/5688#issuecomment-1552067121 and below

Overall I believe this naive implementation is usable already and can serve as a basis for future enhancements.

[1] https://epubs.siam.org/doi/10.1137/04061101X

### CODE REVIEW

**Comment by takagi** [comment]
I'll merge this after the release this week.

File: cupyx/scipy/linalg/_matfuncs.py [hunk 1]
```diff
@@ -35,3 +37,95 @@ def khatri_rao(a, b):
...
+        return cupy.zeros((0, 0), dtype=a.dtype)
+
+    n = a.shape[0]
+
+    # try reducing the norm
+    mu = cupy.diag(a).sum() / n
+    A = a - cupy.eye(n)*mu
+
+    # scale factor
+    nrmA = cupy.linalg.norm(A, ord=1)
```
Comment (takagi) [inline cupyx/scipy/linalg/_matfuncs.py:97]: Only sync here.
```suggestion
    nrmA = cupy.linalg.norm(A, ord=1).item()
```

File: cupyx/scipy/linalg/_matfuncs.py [hunk 2]
```diff
@@ -35,3 +37,95 @@ def khatri_rao(a, b):
...
+
+th13 = 5.37
+
+
+def expm(a):
+    """Compute the matrix exponential.
+
+    Parameters
+    ----------
+    a : ndarray, 2D, float64
```
Comment (takagi) [inline cupyx/scipy/linalg/_matfuncs.py:66]:
```suggestion
    a : ndarray, 2D
```

File: cupyx/scipy/linalg/_matfuncs.py [hunk 3]
```diff
@@ -35,3 +37,95 @@ def khatri_rao(a, b):
...
+    matrix exponential of `a`
+
+    Notes
+    -----
+    Uses (a simplified) version of Algorithm 2.3 of [1]_:
+    a [13 / 13] Pade approximant with scaling and squaring.
+
+    Simplifications:
+    - we always use a [13/13] approximant
+    - no matrix balancing
```
Comment (takagi) [inline cupyx/scipy/linalg/_matfuncs.py:79]:
```suggestion
    Simplifications:
    
        * we always use a [13/13] approximate
        * no matrix balancing
```

File: tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py [hunk 4]
```diff
@@ -66,3 +74,32 @@ def test_equality_of_two_equations(self, dtype):
...
+        a = cupy.array([[0., 0], [0, 0]])
+        assert cupy.abs(cx_linalg.expm(a) - cupy.eye(2)).all() < 1e-10
+
+    def test_empty_matrix_input(self):
+        # handle gh-11082
+        A = np.zeros((0, 0))
+        result = cx_linalg.expm(A)
+        assert result.size == 0
+
+    @testing.numpy_cupy_allclose(scipy_name='scp')
```
Comment (ev-br) [inline tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:91]:
```suggestion
    @testing.numpy_cupy_allclose(scipy_name='scp', contiguous_check=False)
```

File: tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py [hunk 5]
```diff
@@ -66,3 +74,32 @@ def test_equality_of_two_equations(self, dtype):
...
+        a = xp.array([[1, 4], [1, 1]])
+        return scp.linalg.expm(a)
+
+    @pytest.mark.parametrize('a', ([[1, 4], [1, 1]],
+                                   [[1, 3], [1, -1]],
+                                   [[1, 3], [4, 5]],
+                                   [[1, 3], [5, 3]],
+                                   [[4, 5], [-3, -4]])
+                             )
+    @testing.numpy_cupy_allclose(scipy_name='scp')
```
Comment (ev-br) [inline tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:102]:
```suggestion
    @testing.numpy_cupy_allclose(scipy_name='scp', contiguous_check=False)
```

**Review by ev-br — COMMENTED** [review]
Two small tweaks to tests to fix CI failures.
(FWIW, I doubt it's a big deal for 2x2 matrices)

File: cupyx/scipy/linalg/_matfuncs.py [hunk 6]
```diff
@@ -35,3 +37,95 @@ def khatri_rao(a, b):
...
+    A6 = A2 @ A4
+
+    E = cupy.eye(A.shape[0])
+
+    u = A6 @ (b[13]*A6 + b[11]*A4 + b[9]*A2) + \
+        b[7]*A6 + b[5]*A4 + b[3]*A2 + b[1]*E
+    u = A @ u
+
+    v = A6 @ (b[12]*A6 + b[10]*A4 + b[8]*A) + \
+        b[6]*A6 + b[4]*A4 + b[2]*A2 + b[0]*E
```
Comment (leofang) [inline cupyx/scipy/linalg/_matfuncs.py:119]: Can we at least fuse `u` and `v` using `@cupy.fuse()`? Seems quite a bit of kernels launched here.

File: tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py (hunk 4 above)
Comment (takagi) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:91]: Looks good to me.

File: tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py (hunk 5 above)
Comment (takagi) [reply tests/cupyx_tests/scipy_tests/linalg_tests/test_matfuncs.py:102]: ditto.

**Comment by ev-br** [comment]
Committed the review suggestions, including `cupy.fuse` (thanks!). 
 
Performance-wise relative to scipy, the break-even matrix size seems to be in the 100s range; for larger sizes the speed-up is about 2x, and then saturates for still larger sizes (the quick-and-dirty benchmark script is under the fold):

```
$ python cupyx/scipy/signal/bench_expm.py 
n = 10  t_cupy = 0.010493308790028095  t_numpy =3.1411825004033745e-05
n = 50  t_cupy = 0.0015397735312581062  t_numpy =0.0002879052673932165
n = 100  t_cupy = 0.0026254084100946783  t_numpy =0.0037681848817970602
n = 200  t_cupy = 0.0045928616286255415  t_numpy =0.009286615341412835
n = 500  t_cupy = 0.02898122939048335  t_numpy =0.06857758777919225
n = 1000  t_cupy = 0.17829166719922795  t_numpy =0.1998201340759173
```

While this is not much, the path to larger perf gains is reasonably clear and I'd suggest to keep it in the list of future improvements.

<details>

```
import numpy as np
import cupy
from cupyx.scipy.linalg import expm
import scipy.linalg as sl

import timeit

def time(n):
    np.random.seed(12345)
    a = np.random.uniform(size=(n, n))

    nreps = 10000
    t_numpy = timeit.timeit(setup='import scipy.linalg as sl',
                            stmt='sl.expm(a)', globals=locals(), number=nreps)
    t_numpy /= nreps

    an = cupy.asarray(a)
    nreps = 100
    t_cupy = timeit.timeit(setup='from cupyx.scipy.linalg import expm',
                           stmt='expm(an)', globals=locals(), number=nreps)
    t_cupy /= nreps
    return t_numpy, t_cupy


for n in [10, 50, 100, 200, 500, 1000]:
    t_numpy, t_cupy = time(n)
    print(f'{n = }  {t_cupy = }  {t_numpy =}')
```
</details>

**Comment by chiwhalee** [comment]
Hi, @ev-br, thank you for your work. There is another function expm_multiply in scipy.  Would you consider  implementing this function in future?  Can the present expm be used in place of expm_multiply (to some extent)?

**Comment by ev-br** [comment]
You mean https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.expm_multiply.html for sparse matrices? 

I personally prefer to scope it for after the dense version, are interested in working on it in the mean time?

**Review by takagi — APPROVED** [review]
LGTM!

**Comment by takagi** [comment]
Thanks, @ev-br! I made an issue about expm performance improvement at #7769.

**Comment by chiwhalee** [comment]
> You mean https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.expm_multiply.html for sparse matrices?
> 
> I personally prefer to scope it for after the dense version, are you interested in working on it in the mean time?

@ev-br Sorry for the delayed response. I have no experience with developing module for cupy. Would it work if I translate the scipy version of expm_multiply line by line to cupy （with minor other changes)？

**Comment by ev-br** [comment]
Sadly, I don't believe simply translating scipy sparse linalg code is going to work. A better bet is to read the al-mohy & higham paper linked from the docstring and implement the algorithm. Maybe the scipy code is helpful, too.

**Pull request merged**